File: packages/files/python/src/immukv_files/_internal/hashing.py
"""SHA-256 streaming hash computation for file content."""

import hashlib
from pathlib import Path
from typing import AsyncIterator, BinaryIO, Iterator, Tuple, TypeVar, Union

from immukv_files.types import ContentHash
from immukv_files._internal.types import content_hash_from_digest

K = TypeVar("K", bound=str)


def compute_hash_from_bytes(data: bytes) -> ContentHash[K]:
    """Compute SHA-256 hash of bytes.

    Args:
        data: Bytes to hash

    Returns:
        ContentHash with 'sha256:' prefix
    """
    hash_obj = hashlib.sha256()
    hash_obj.update(data)
    hex_digest = hash_obj.hexdigest()
    return content_hash_from_digest(hex_digest)


def compute_hash_from_file(file_obj: BinaryIO) -> Tuple[ContentHash[K], bytes, int]:
    """Compute SHA-256 hash of a file object.

    Reads the entire file and buffers it for upload.

    Args:
        file_obj: Binary file object to hash

    Returns:
        Tuple of (content_hash, buffer, content_length)
    """
    hash_obj = hashlib.sha256()
    chunks: list[bytes] = []

    while True:
        chunk = file_obj.read(65536)  # 64KB chunks
        if not chunk:
            break
        hash_obj.update(chunk)
        chunks.append(chunk)

    buffer = b"".join(chunks)
    hex_digest = hash_obj.hexdigest()
    return content_hash_from_digest(hex_digest), buffer, len(buffer)


def compute_hash_from_path(file_path: Union[str, Path]) -> Tuple[ContentHash[K], bytes, int]:
    """Compute SHA-256 hash of a file at the given path.

    Reads the entire file and buffers it for upload.

    Args:
        file_path: Path to file to hash

    Returns:
        Tuple of (content_hash, buffer, content_length)
    """
    path = Path(file_path)
    with path.open("rb") as f:
        return compute_hash_from_file(f)


def compute_hash_from_iterator(data: Iterator[bytes]) -> Tuple[ContentHash[K], bytes, int]:
    """Compute SHA-256 hash from an iterator of bytes.

    Consumes the iterator and buffers content for upload.

    Args:
        data: Iterator of bytes chunks

    Returns:
        Tuple of (content_hash, buffer, content_length)
    """
    hash_obj = hashlib.sha256()
    chunks: list[bytes] = []

    for chunk in data:
        hash_obj.update(chunk)
        chunks.append(chunk)

    buffer = b"".join(chunks)
    hex_digest = hash_obj.hexdigest()
    return content_hash_from_digest(hex_digest), buffer, len(buffer)


def verify_bytes_hash(data: bytes, expected_hash: ContentHash[K]) -> bool:
    """Verify that bytes match an expected content hash.

    Args:
        data: Bytes to verify
        expected_hash: Expected ContentHash

    Returns:
        True if hash matches
    """
    actual_hash: ContentHash[str] = compute_hash_from_bytes(data)
    return str(actual_hash) == str(expected_hash)


def verify_iterator_hash(data: Iterator[bytes], expected_hash: ContentHash[K]) -> bool:
    """Verify that iterator content matches an expected content hash.

    Args:
        data: Iterator of bytes to verify
        expected_hash: Expected ContentHash

    Returns:
        True if hash matches
    """
    actual_hash: ContentHash[str]
    actual_hash, _, _ = compute_hash_from_iterator(data)
    return str(actual_hash) == str(expected_hash)


# =============================================================================
# Async hash functions
# =============================================================================


async def compute_hash_from_async_iterator(
    data: AsyncIterator[bytes],
) -> Tuple[ContentHash[str], bytes, int]:
    """Compute SHA-256 hash from async iterator.

    Consumes the iterator and buffers content.

    Args:
        data: Async iterator of byte chunks

    Returns:
        Tuple of (content_hash, buffer, content_length)
    """
    hash_obj = hashlib.sha256()
    chunks: list[bytes] = []

    async for chunk in data:
        hash_obj.update(chunk)
        chunks.append(chunk)

    buffer = b"".join(chunks)
    return content_hash_from_digest(hash_obj.hexdigest()), buffer, len(buffer)


async def verify_async_iterator_hash(
    data: AsyncIterator[bytes], expected_hash: ContentHash[K]
) -> bool:
    """Verify that async iterator content matches an expected content hash.

    Args:
        data: Async iterator of bytes to verify
        expected_hash: Expected ContentHash

    Returns:
        True if hash matches
    """
    actual_hash: ContentHash[str]
    actual_hash, _, _ = await compute_hash_from_async_iterator(data)
    return str(actual_hash) == str(expected_hash)

File: packages/files/typescript/src/internal/hashing.ts
/**
 * SHA-256 streaming hash computation for file content.
 */

import * as crypto from "crypto";
import type { Readable } from "stream";
import type { ContentHash } from "../types";
import { contentHashFromDigest } from "./types";

/**
 * Compute SHA-256 hash of a buffer.
 *
 * @param data - Buffer to hash
 * @returns ContentHash with 'sha256:' prefix
 */
export function computeHashFromBuffer<K extends string>(
  data: Buffer,
): ContentHash<K> {
  const hash = crypto.createHash("sha256");
  hash.update(data);
  const hexDigest = hash.digest("hex");
  return contentHashFromDigest<K>(hexDigest);
}

/**
 * Compute SHA-256 hash of a string.
 *
 * @param data - String to hash (encoded as UTF-8)
 * @returns ContentHash with 'sha256:' prefix
 */
export function computeHashFromString<K extends string>(
  data: string,
): ContentHash<K> {
  return computeHashFromBuffer<K>(Buffer.from(data, "utf-8"));
}

/**
 * Result of streaming hash computation.
 * Contains both the hash and the buffered content for upload.
 */
export interface StreamHashResult<K extends string> {
  /** SHA-256 hash of the content */
  contentHash: ContentHash<K>;

  /** Buffered content for upload */
  buffer: Buffer;

  /** Content length in bytes */
  contentLength: number;
}

/**
 * Compute SHA-256 hash of a readable stream.
 *
 * Buffers the entire stream to compute hash and return content for upload.
 * For large files, consider using multipart upload with per-part hashing.
 *
 * @param stream - Readable stream to hash
 * @returns Promise resolving to hash and buffered content
 */
export async function computeHashFromStream<K extends string>(
  stream: Readable,
): Promise<StreamHashResult<K>> {
  return new Promise((resolve, reject) => {
    const hash = crypto.createHash("sha256");
    const chunks: Buffer[] = [];

    stream.on("data", (chunk: Buffer | string) => {
      const buffer = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);
      hash.update(buffer);
      chunks.push(buffer);
    });

    stream.on("end", () => {
      const hexDigest = hash.digest("hex");
      const buffer = Buffer.concat(chunks);
      resolve({
        contentHash: contentHashFromDigest<K>(hexDigest),
        buffer,
        contentLength: buffer.length,
      });
    });

    stream.on("error", (error) => {
      reject(error);
    });
  });
}

/**
 * Verify that a buffer matches an expected content hash.
 *
 * @param data - Buffer to verify
 * @param expectedHash - Expected ContentHash
 * @returns true if hash matches
 */
export function verifyBufferHash<K extends string>(
  data: Buffer,
  expectedHash: ContentHash<K>,
): boolean {
  const actualHash = computeHashFromBuffer<K>(data);
  return actualHash === expectedHash;
}

/**
 * Verify that a stream matches an expected content hash.
 *
 * @param stream - Readable stream to verify
 * @param expectedHash - Expected ContentHash
 * @returns Promise resolving to true if hash matches
 */
export async function verifyStreamHash<K extends string>(
  stream: Readable,
  expectedHash: ContentHash<K>,
): Promise<boolean> {
  const result = await computeHashFromStream<K>(stream);
  return result.contentHash === expectedHash;
}
